### PR TITLE
Create rliable.py

### PR DESCRIPTION
**Pull request by Rahul-Choudhary-3614**

### CODE REVIEW

File: hive/utils/rliable.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+from scipy.stats.stats import find_repeats
+import matplotlib.patches as mpatches
+from matplotlib.ticker import MaxNLocator
+import argparse
+
+#Aggregate metrics
+IQM = lambda x: metrics.aggregate_iqm(x) # Interquartile Mean
+OG = lambda x: metrics.aggregate_optimality_gap(x, 1.0) # Optimality Gap
+MEAN = lambda x: metrics.aggregate_mean(x)
+MEDIAN = lambda x: metrics.aggregate_median(x)
```
Comment (dapatil211) [inline hive/utils/rliable.py]: There's no reason to put these outside. You can just use the functions directly

File: hive/utils/rliable.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
+from rliable import library as rly
+from rliable import metrics
+from rliable import plot_utils
+import matplotlib.pyplot as plt
+import numpy as np
+import seaborn as sns
+from scipy.stats.stats import find_repeats
+import matplotlib.patches as mpatches
+from matplotlib.ticker import MaxNLocator
+import argparse
```
Comment (dapatil211) [inline hive/utils/rliable.py]: Sort imports with isort (your editor likely already has functionality for this)

Comment (dapatil211) [inline hive/utils/rliable.py]: Sort imports with isort (your editor likely already has functionality for this)

File: hive/utils/rliable.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+
+def save_fig(fig, name):
+  file_name = '{}.pdf'.format(name)
+  fig.savefig(file_name, format='pdf', bbox_inches='tight')
+
+def convert_to_matrix(score_dict):
+   keys = sorted(list(score_dict.keys()))
+   return np.stack([score_dict[k] for k in keys], axis=1)
+
+def load_and_read_rewards(env_names,path,min_env_score,max_env_score,Normalization):
```
Comment (dapatil211) [inline hive/utils/rliable.py]: normalization should be lowercase

File: hive/utils/rliable.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+  score_dict = {}
+  for env_index,env_name in enumerate(env_names):
+    score_ = np.load(path+'/'+env_name+'/Rewards.npy')
```
Comment (dapatil211) [inline hive/utils/rliable.py]: 1. Use os.path.join
2. Where are you getting this file name from? You should be loading the rewards from the chomplogger files.

File: hive/utils/rliable.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+    if Normalization:
+      score_dict[env_name] = score_normalization(score_, min_env_score[env_index], max_env_score[env_index])
```
Comment (dapatil211) [inline hive/utils/rliable.py]: min_env_score is a scalar isn't it?

File: hive/utils/rliable.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+  for agent_name in agent_names:
+    path_temp = path+'/'+agent_name
+    score_data_dict[agent_name] = load_and_read_rewards(env_names,path_temp,min_env_score,max_env_score,Normalization)
+
+  #Aggregates
+  aggregates_score_dict = {key: val[:num_runs_aggregates] for key, val in score_data_dict.items()}
+  aggregate_func = lambda x: np.array([MEDIAN(x), IQM(x), MEAN(x), OG(x)])
+  aggregate_scores, aggregate_interval_estimates = rly.get_interval_estimates(aggregates_score_dict, aggregate_func, reps=50000)
+
+  if plot:
```
Comment (dapatil211) [inline hive/utils/rliable.py]: Hold off on this for now. We can add it once we have a good way of getting the raw metrics

**Review by dapatil211 — CHANGES_REQUESTED** [review]
Please use black formatter.

File: hive/utils/rliable_matrix.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    reps = 1000
+    probabilities, probability_cis = rly.get_interval_estimates(
+        all_pairs, metrics.probability_of_improvement, reps=reps
+    )
+
+    # print('Aggregate Scores {}'.format(aggregate_scores))
+    # print('Aggregate Interval Estimates {}'.format(aggregate_interval_estimates))
+    # print('Score Distributions {}'.format(score_distributions))
+    # print('Average Score Distributions {}'.format(avg_score_distributions))
+    # print('Probabilities {}'.format(probabilities))
```
Comment (dapatil211) [inline hive/utils/rliable_matrix.py:148]: Delete commented code

File: hive/utils/rliable_matrix.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        ]
+    )
+    aggregate_scores, aggregate_interval_estimates = rly.get_interval_estimates(
+        aggregates_score_dict, aggregate_func, reps=50000
+    )
+
+    # Calculate score distributions and average score distributions
+    score_dict = {key: score_data_dict[key][:10] for key in agent_names}
+    TAU = np.linspace(0.0, 2.0, 201)
+    # Higher value of reps corresponds to more accurate estimates but are slower to computed. `reps` corresponds to number of bootstrap resamples.
```
Comment (dapatil211) [inline hive/utils/rliable_matrix.py:116]: longer than 88 columns

File: hive/utils/rliable_matrix.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+              num_runs_aggregates(int): Number of runs to be considered for aggregation.
+              path: Path to the folder  where the rewards/scores are stored.
+              output_path: Path to the folder where the matrices are stored.
+
+            normalization (bool): Indicates whether the normalisation is applied or not.
+            min_env_score: List of minimum value if normalisation is applied. Value must be provided for each environment.
+                          If single float value is provided then it is taken as minimum value for all environment.
+            max_env_score: List of maximum value if normalisation is applied. Value must be provided for each environment.
+                          If single float value is provided then it is taken as maximum value for all environment
+    """
```
Comment (dapatil211) [inline hive/utils/rliable_matrix.py:81]: Longer than 88 columns

File: hive/utils/rliable_matrix.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+            job_files = [
+                i for i in os.listdir(os.path.join(path_temp, seed)) if i[0] != "."
+            ]
+            for job_file in job_files:
+                pkl_file = open(
+                    os.path.join(
+                        path_temp,
+                        seed,
+                        job_file,
+                        "logger/logger_0/log_data.p",
```
Comment (dapatil211) [inline hive/utils/rliable_matrix.py:42]: You should not assume you have this file structure. For now, write the functions in the following way:
`process_rewards(rewards)`, where rewards is whatever format/structure you want (ie assume the data has already been loaded for you. Then write a separate function to read the rewards. For this, you can use the functions in `visualization.py` if that's easier (eg something like `find_all_experiments_data()` and you just write code to pass the appropriate arguments.

**Review by dapatil211 — CHANGES_REQUESTED** [review]

**Comment by Prishruit** [comment]
@Rahul-Choudhary-3614 could you update the code so that it is consistent with the current base branch.

**Pull request closed without merging**